### PR TITLE
Fix `/share` and cleanup and reorganize frontend locale loading

### PR DESCRIPTION
**Pull request by renchap**
My react-intl upgrade PR broke the locale loading on `/share`. This was caused by `getLocale` being called when importing files, as it is called in the root scope of the file.

This is a bad practice and can easily lead to those files being imported/loaded while the locale is not loaded yet. The current code needs to dynamically import the various React components after `loadLocale` has been called, which is error-prone.

Now, locale data is loaded when the component/code is called, not imported. This avoid the need to dynamically import this code once the locale data has been loaded

While I was working on this, I made some additional changes that I did not wanted to do in my initial PR to not make it too complex:
- Removed some unused files I missed in my last commit
- No more locale loading as a side-effect
- No need for the `locale` prop anymore for root React components, this is read from `<html lang>`
- Converted locale-related files to Typescript
- locale-related TS code is in `mastodon/locales/`, which only exports the needed functions & types

### CODE REVIEW

**Comment by renchap** [comment]
Note: I can also rework this to not move the locale JSON files and have the TS code in another directory if you prefer (`mastodon/intl`?)

File: .prettierignore [hunk 1]
```diff
@@ -61,7 +61,7 @@ docker-compose.override.yml
 /app/javascript/mastodon/features/emoji/emoji_map.json
 
 # Ignore locale files
-/app/javascript/mastodon/locales
+/app/javascript/mastodon/locales/data/*.json
```
Comment (nschonni) [inline .prettierignore:64]: It seems like the JSON files don't need to be ignored anymore after the migration from `translationRunner.js`

**Comment by renchap** [comment]
> I understand the reason for moving the JSON files and I agree with it, but I'm not sure how well Crowdin handles that, so I'd like that to be postponed until we make sure this won't break anything with Crowdin.

I can leave the JSON in `locales/` and move the TS files elsewhere, but I am not sure where. Ideas:
- `mastodon/intl/`
- `mastodon/locale/` (without an `s`)
- `mastodon/utils/intl/`

**Comment by ClearlyClaire** [comment]
Do the TS files need to be kept out of that folder as a first step? If not, I'd postpone that split to a later PR, if yes, I'm fine with `mastodon/intl`

File: .prettierignore (hunk 1 above)
Comment (renchap) [reply .prettierignore:64]: They are automatically generated, so we dont want Prettier to process them

**Comment by renchap** [comment]
I removed the JSON file renaming.

All TS files are in `mastodon/locales/` with the JSON files in the same directory. I split them in multiple files, `index.ts` re-exports what should be used in the app.

**Review by ClearlyClaire — APPROVED** [review]

**Pull request merged**